apps/ledger-live-desktop/src/renderer/components/CustomImage/ImageCropper.tsx

```diff
@@ -0,0 +1,244 @@
+import React, { useCallback, useEffect, useRef, useState } from "react";
+import { ImageBase64Data, ImageDimensions } from "./types";
+import { useDebounce } from "@ledgerhq/live-common/hooks/useDebounce";
+import { Flex, Icons } from "@ledgerhq/react-ui";
+import Cropper, { Area, CropperProps } from "react-easy-crop";
+import { createCanvas, getRadianAngle, rotateSize } from "./imageUtils";
+import { ImageCropError } from "@ledgerhq/live-common/customImage/errors";
+
+export type CropResult = ImageDimensions & ImageBase64Data;
+
+type Crop = Area;
+
+const MAX_ZOOM = 10;
+
+export type CropParams = {
+  crop: Crop;
+  rotationIncrements: number;
+  zoom: number;
+  imageUuid: string;
+};
+
+export type Props = ImageBase64Data & {
+  targetDimensions: { width: number; height: number };
+  onResult: (_: CropResult) => void;
+  onError: (_: Error) => void;
+  setCropParams: (_: CropParams) => void;
+  setLoading: (_: boolean) => void;
+  withButton?: boolean;
+  initialCropParams?: CropParams;
+};
+
+async function cropAndResizeImage(
+  image: HTMLImageElement,
+  targetDimensions: ImageDimensions,
+  pixelCrop: Crop,
+  rotationIncrements: number,
+): Promise<CropResult> {
+  if (!image.complete) {
+    await new Promise(resolve => {
+      image.onload = resolve;
```

Comment:
please handle
`image.onerror = reject`

Reply:
more generally speaking it actually would be better that the `loadImage` responsability is handled "before". so you catch any possible load issue.

to me you should have somewhere a `loadImage(src)` that is doing the promise (without need to if image complete)

Reply:
Maybe it would be better if instead of an `HTMLImageElement` I passed a `React.ObjectRef<HTMLImageElement>` to make it more explicit that the image loading is handled here only if needed.  (And i will also add the `image.onerror = reject`)

Here the image component is mounted which is convenient for several reasons: it (re)loads the image automatically every time it receives a new `src` prop, it also has the `onError` prop already. So the `loadImage(src)` you are suggesting is already done implicitly by React.

Then this function only receives the ref to that component so in most cases the loading of the image will already have been triggered.

The point of this logic is to mount (& load) the image just once with React, then the crop result can be computed as many times as needed (whenever the user modifies the cropping parameters through the cropping UI) without having to each time load the image, which proved to be quite laggy for big images.

Reply:
you can do this too, it means you no longer need this promise step, right?

Reply:
I don't know I feel like there could still be a corner case where the image has been loaded in the crop UI component (because it loads the image itself and there we can't access the image ref) but not in our own img component, so the onCropComplete could be triggered before the image is loaded ...